File: src/ios/YoikIBeacon.m
/*
The MIT License (MIT)

Copyright (c) 2014

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */

#import "YoikIBeacon.h"
#import <Cordova/CDVJSON.h>


#pragma mark -
#pragma mark YoikIBeacon

@implementation YoikIBeacon

static int NIGH_PROXIMITY = -30;

@synthesize locationManager;

- (CDVPlugin*)initWithWebView:(UIWebView*)theWebView
{
    if (self) {
        self.locationManager = [[CLLocationManager alloc] init];
        self.locationManager.delegate = self; // Tells the location manager to send updates to this object

        self.lastNigh = [[NSDate alloc] init];
        self.lastFar = [[NSDate alloc] init];

        self.beaconDict = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)addRegion:(CDVInvokedUrlCommand*)command
{
    NSString* callbackId = command.callbackId;

    @try {
        NSArray* arguments = command.arguments;
        NSUUID *uuid = [[NSUUID alloc] initWithUUIDString:[arguments objectAtIndex:1]];
        NSString* identifier = [arguments objectAtIndex:0];
        NSLog(@"added region.. %@", [arguments objectAtIndex:1]);

        CLBeaconRegion *myRegion = [[CLBeaconRegion alloc] initWithProximityUUID:uuid identifier:identifier];

        [self.beaconDict setObject:myRegion forKey:identifier];
        [self.locationManager startMonitoringForRegion: myRegion];

        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:result callbackId:callbackId];
    }
    @catch (NSException * e) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        [self.commandDelegate sendPluginResult:result callbackId:callbackId];
    }
}


#pragma mark -
#pragma mark CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager*)manager didEnterRegion:(CLRegion*)region
{
    NSLog(@"Entered region..%@", region.identifier);
    [self.locationManager startRangingBeaconsInRegion: self.beaconDict[region.identifier]];

    NSMutableDictionary *inner = [[NSMutableDictionary alloc] init];
    [inner setObject:region.identifier forKey:@"identifier"];
    NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
    [result setObject:inner forKey:@"ibeacon"];

    NSString *jsStatement = [NSString stringWithFormat:@"cordova.fireDocumentEvent('ibeaconenter', %@);", [result JSONString]];
    [self.commandDelegate evalJs:jsStatement];
}

-(void)locationManager:(CLLocationManager*)manager didExitRegion:(CLRegion*)region
{
    NSLog(@"Exited region..%@", region.identifier);
    [self.locationManager stopRangingBeaconsInRegion: self.beaconDict[region.identifier]];

    NSMutableDictionary *inner = [[NSMutableDictionary alloc] init];
    [inner setObject:region.identifier forKey:@"identifier"];
    NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
    [result setObject:inner forKey:@"ibeacon"];

    NSString *jsStatement = [NSString stringWithFormat:@"cordova.fireDocumentEvent('ibeaconexit', %@);", [result JSONString]];
    [self.commandDelegate evalJs:jsStatement];
}

-(void)locationManager:(CLLocationManager*)manager
       didRangeBeacons:(NSArray*)beacons
              inRegion:(CLBeaconRegion*)region
{
    // Beacon found!
    if (beacons.count > 0) {

        CLBeacon *foundBeacon = [beacons firstObject];

        if (foundBeacon.rssi >= NIGH_PROXIMITY && foundBeacon.rssi < 0) {
            NSLog(@"%ld - %ld", (long)foundBeacon.rssi, (long)foundBeacon.major);

            NSTimeInterval secs = [self.lastNigh timeIntervalSinceNow];

            if (secs < -6) {
                [self sendIbeaconEvent:foundBeacon forRegion:region forRange:@"nigh"];

                self.lastNigh = [[NSDate alloc] init];
            }

        } else {
            switch (foundBeacon.proximity) {
                case CLProximityNear:
                case CLProximityFar:
                {
                    NSTimeInterval secs = [self.lastFar timeIntervalSinceNow];

                    if (secs < -60) {
                        [self sendIbeaconEvent:foundBeacon forRegion:region forRange:[self regionText:foundBeacon]];

                        self.lastFar = [[NSDate alloc] init];
                    }
                }
                    break;
                default:
                    break;
            }
        }
    }

}

- (void)sendIbeaconEvent:(CLBeacon *)foundBeacon forRegion:(CLRegion *) region forRange:(NSString *) range
{
    NSLog(@"Sending event");
    // You can retrieve the beacon data from its properties
    NSString *uuid = foundBeacon.proximityUUID.UUIDString;
    NSString *major = [NSString stringWithFormat:@"%@", foundBeacon.major];
    NSString *minor = [NSString stringWithFormat:@"%@", foundBeacon.minor];

    NSMutableDictionary *inner = [[NSMutableDictionary alloc] init];

    [inner setObject: [uuid lowercaseString] forKey:@"uuid"];
    [inner setObject: major forKey:@"major"];
    [inner setObject: minor forKey:@"minor"];
    [inner setObject: range forKey:@"range"];
    [inner setObject:region.identifier forKey:@"identifier"];

    NSMutableDictionary *result = [[NSMutableDictionary alloc] init];
    [result setObject:inner forKey:@"ibeacon"];

    NSLog(@"%@", [result JSONString]);

    NSString *jsStatement = [NSString stringWithFormat:@"cordova.fireDocumentEvent('ibeacon', %@);", [result JSONString]];

    [self.commandDelegate evalJs:jsStatement];
}


- (void)locationManager:(CLLocationManager *)manager didStartMonitoringForRegion:(CLRegion *)region
{
    NSLog(@"hello %@", region.identifier);
    [self.locationManager startRangingBeaconsInRegion: self.beaconDict[region.identifier]];
}

- (void)locationManager: (CLLocationManager *)manager
       didFailWithError: (NSError *)error
{
    [manager stopUpdatingLocation];
    NSLog(@"error%@",error);
}

- (void)dealloc
{
    self.locationManager.delegate = nil;
    self.beaconDict = nil;
}

- (void)onReset
{
    self.beaconDict = nil;
}

- (NSString *)regionText:(CLBeacon *)beacon
{
    switch (beacon.proximity) {
        case CLProximityFar:
            return @"far";
        case CLProximityImmediate:
            return @"immediate";
        case CLProximityNear:
            return @"near";
        default:
            return @"unknown";
    }
}

@end

File: src/android/net/yoik/cordova/plugins/ibeacon/YoikIBeacon.java
/*
The MIT License (MIT)

Copyright (c) 2014

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */
package net.yoik.cordova.plugins.ibeacon;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.app.Activity;
import android.content.Intent;
import android.content.Context;
import android.util.Log;
import android.os.RemoteException;
import android.content.ServiceConnection;
import android.text.format.Time;

import java.util.Collection;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView;

import com.radiusnetworks.ibeacon.IBeacon;
import com.radiusnetworks.ibeacon.IBeaconData;
import com.radiusnetworks.ibeacon.IBeaconConsumer;
import com.radiusnetworks.ibeacon.IBeaconDataNotifier;
import com.radiusnetworks.ibeacon.IBeaconManager;
import com.radiusnetworks.ibeacon.client.DataProviderException;
import com.radiusnetworks.ibeacon.MonitorNotifier;
import com.radiusnetworks.ibeacon.Region;
import com.radiusnetworks.ibeacon.RangeNotifier;

/**
 * This calls out to the ZXing barcode reader and returns the result.
 *
 * @sa https://github.com/apache/cordova-android/blob/master/framework/src/org/apache/cordova/CordovaPlugin.java
 */
public class YoikIBeacon extends CordovaPlugin implements IBeaconConsumer, MonitorNotifier, RangeNotifier, IBeaconDataNotifier {
    public static final int REQUEST_CODE = 0x0ba7c0de;

    private static final String TAG = "YoikIBeacon";

    private static final String ACTION_ADDREGION = "addRegion";

    private static final int NEAR_FAR_FREQUENCY = 1 * 60 * 1000;
    private static final int NIGH_RSSI = -30;
    private static final int NIGH_FREQUENCY = 6 * 1000;
    private static final int PROXIMITY_NIGH = 100;

    private IBeaconManager iBeaconManager;

    private Time lastNigh;
    private Time lastFar;

    private Boolean firstNearFar;

    private CallbackContext callbackContext;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);

        final Activity activity = cordova.getActivity();
        final YoikIBeacon that = this;

        this.lastNigh = new Time();
        this.lastNigh.setToNow();

        this.lastFar = new Time();
        this.lastFar.setToNow();
        this.firstNearFar = true;

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {

                iBeaconManager = IBeaconManager.getInstanceForApplication(activity);
                iBeaconManager.bind(that);
            }
        });
    }

    /**
     * Executes the request.
     *
     * This method is called from the WebView thread. To do a non-trivial amount of work, use:
     *     cordova.getThreadPool().execute(runnable);
     *
     * To run on the UI thread, use:
     *     cordova.getActivity().runOnUiThread(runnable);
     *
     * @param action          The action to execute.
     * @param data            The exec() arguments.
     * @param callbackContext The callback context used when calling back into JavaScript.
     * @return                Whether the action was valid.
     *
     * @sa https://github.com/apache/cordova-android/blob/master/framework/src/org/apache/cordova/CordovaPlugin.java
     */
    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) {
        this.callbackContext = callbackContext;

        Log.d(TAG, "execute " + action);

        if (action.equals(ACTION_ADDREGION)) {
            addRegion(data, callbackContext);
        } else if (action.equals("anotheraction")) {

        } else {
            return false;
        }
        return true;
    }

    private void addRegion(JSONArray data, CallbackContext callbackContext) {

        final JSONArray data2 = data;
        final CallbackContext callbackContext2 = callbackContext;

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    iBeaconManager.startMonitoringBeaconsInRegion(new Region(data2.optString(0), data2.optString(1), null, null));
                    callbackContext2.success();

                } catch (RemoteException e) {
                    callbackContext2.error("Could not add region");
                }
            }
        });

    }

    private void init(CallbackContext callbackContext) {
        Log.d(TAG, "Enabling plugin");

        callbackContext.success();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        iBeaconManager.unBind(this);
    }

    @Override
    public void onPause(boolean multitasking) {
        super.onPause(multitasking);
        if (iBeaconManager.isBound(this)) iBeaconManager.setBackgroundMode(this, true);
    }

    @Override
    public void onResume(boolean multitasking) {
        super.onResume(multitasking);
        if (iBeaconManager.isBound(this)) iBeaconManager.setBackgroundMode(this, false);
    }

    public Context getApplicationContext() {
        return cordova.getActivity().getApplicationContext();
    }

    public void unbindService(ServiceConnection connection) {
        cordova.getActivity().unbindService(connection);
    }

    public boolean bindService(Intent intent, ServiceConnection connection, int mode) {
        return cordova.getActivity().bindService(intent, connection, mode);
    }

    @Override
    public void onIBeaconServiceConnect() {
        iBeaconManager.setMonitorNotifier(this);
    }

    @Override
    public void didEnterRegion(Region region) {
        Log.d(TAG, "I just saw an iBeacon for the first time!");

        try {
            iBeaconManager.startRangingBeaconsInRegion(region);
            iBeaconManager.setRangeNotifier(this);

            JSONObject obj = new JSONObject();
            obj.put("identifier", region.getUniqueId());

            JSONObject result = new JSONObject();
            result.put("ibeacon", obj);

            final String jsStatement = String.format("cordova.fireDocumentEvent('ibeaconenter', %s);", result.toString());

            cordova.getActivity().runOnUiThread(
                new Runnable() {
                    @Override
                     public void run() {
                         webView.loadUrl("javascript:" + jsStatement);
                     }
                }
            );

        } catch (JSONException e) {
            e.printStackTrace();
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void didExitRegion(Region region) {
        Log.d(TAG, "I no longer see an iBeacon");

        try {
            iBeaconManager.startRangingBeaconsInRegion(region);

            JSONObject obj = new JSONObject();
            obj.put("identifier", region.getUniqueId());

            JSONObject result = new JSONObject();
            result.put("ibeacon", obj);

            final String jsStatement = String.format("cordova.fireDocumentEvent('ibeaconexit', %s);", result.toString());

            cordova.getActivity().runOnUiThread(
                new Runnable() {
                    @Override
                     public void run() {
                         webView.loadUrl("javascript:" + jsStatement);
                     }
                }
            );

        } catch (JSONException e) {
            e.printStackTrace();
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void didDetermineStateForRegion(int state, Region region) {
        // logToDisplay("I have just switched from seeing/not seeing iBeacons: "+state);
        Log.d(TAG, "I have just switched from seeing/not seeing iBeacons: "+state);
    }

    @Override
    public void didRangeBeaconsInRegion(Collection<IBeacon> iBeacons, Region region) {

        for (IBeacon iBeacon: iBeacons) {
            Integer proximity = iBeacon.getProximity();
            Integer rssi = iBeacon.getRssi();

            // custom check for nigh proximity,
            if (rssi > NIGH_RSSI && rssi < 0) {
                Log.d(TAG, "Found One: " + rssi + " " + iBeacon.getProximityUuid() + " " + iBeacon.getMajor() + " " + iBeacon.getMinor());

                Time now = new Time();
                now.setToNow();

                if ((now.toMillis(false) - this.lastNigh.toMillis(false)) > NIGH_FREQUENCY) {
                    sendIbeaconEvent(iBeacon, region, PROXIMITY_NIGH);
                    this.lastNigh.setToNow();
                }

            } else if (proximity == IBeacon.PROXIMITY_FAR || proximity == IBeacon.PROXIMITY_NEAR) {

                Time now = new Time();
                now.setToNow();

                if (this.firstNearFar || (now.toMillis(false) - this.lastFar.toMillis(false)) > NEAR_FAR_FREQUENCY) {
                    Log.d(TAG, "Found One Near/Far: "+iBeacon.getAccuracy());

                    sendIbeaconEvent(iBeacon, region, proximity);
                    this.lastFar.setToNow();
                    this.firstNearFar = false;
                }
            }
        }

    }

    @Override
    public void iBeaconDataUpdate(IBeacon iBeacon, IBeaconData iBeaconData, DataProviderException e) {
        if (e != null) {
            Log.d(TAG, "data fetch error:"+e);
        }
        if (iBeaconData != null) {
            String displayString = iBeacon.getProximityUuid()+" "+iBeacon.getMajor()+" "+iBeacon.getMinor()+"\n"+"Welcome message:"+iBeaconData.get("welcomeMessage");

        }
    }

    private void sendIbeaconEvent(IBeacon iBeacon, Region region, Integer range)
    {
        try {
            Log.d(TAG, "Firing Event");

            JSONObject obj = new JSONObject();
            obj.put("uuid", iBeacon.getProximityUuid());
            obj.put("major", iBeacon.getMajor());
            obj.put("minor", iBeacon.getMinor());
            obj.put("range", proximityText(range));
            obj.put("identifier", region.getUniqueId());

            JSONObject result = new JSONObject();
            result.put("ibeacon", obj);

            final String jsStatement = String.format("cordova.fireDocumentEvent('ibeacon', %s);", result.toString());

            cordova.getActivity().runOnUiThread(
                new Runnable() {
                    @Override
                     public void run() {
                         webView.loadUrl("javascript:" + jsStatement);
                     }
                }
            );

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private String proximityText(Integer proximity) {
        switch (proximity) {
            case PROXIMITY_NIGH:
                return "nigh";
            case IBeacon.PROXIMITY_NEAR:
                return "near";
            case IBeacon.PROXIMITY_FAR:
                return "far";
            case IBeacon.PROXIMITY_IMMEDIATE:
                return "immediate";
            default:
                return "unknown";
        }
    }

}